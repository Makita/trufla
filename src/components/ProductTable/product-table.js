/* eslint-disable no-magic-numbers */
// @flow
import React, { Component } from 'react';

import { Table } from 'react-bootstrap';

const productsPerPage = 20;

const ProductRow = ({
  id,
  name,
  price,
  department,
  promoCode
}: {
  id: number,
  name: string,
  price: number,
  department: Object,
  promoCode: Object
}) => {
  return (
    <tr>
      <td>{id}</td>
      <td>{name}</td>
      <td>{price}</td>
      <td>{department}</td>
      <td>{promoCode}</td>
    </tr>
  )
}

const ProductRows = ({ products, page }) => {
  const keys = Array.from(
    Array(productsPerPage).keys(),
    // eslint-disable-next-line id-length
    x => x + (productsPerPage * (page - 1)) + 1
  );

  return keys.map((key) => {
    const product = products[key];

    if (typeof product === "undefined" || typeof product["promo-code"] === "undefined") return null;

    const code = product["promo-code"].code === "nopromo" ? "" : product["promo-code"].code;

    return <ProductRow
      id={product.id}
      name={product.name}
      price={product.price}
      department={product.department.name}
      promoCode={code}
    />;
  });
}

export default class ProductTable extends Component<{ page: number }, {}> {
  state = {
    products: {},
  }

  componentDidMount() {
    fetch('/products/1/21')
      .then(res => res.json())
      .then(json => this.handleJSON(json.data));
  }

  handleJSON(data) {
    // Take each product and add it to an object structured like the state
    const newValues = data.reduce((accumulator, value) => {
      return {
        ...accumulator,
        [value.id]: value.attributes
      };
    }, {})

    // eslint-disable-next-line no-unused-vars
    this.setState({
      products: newValues
    });
  }

  render() {
    const { products } = this.state;
    const { page } = this.props;

    return (
      <Table striped responsive hover>
        <thead>
          <tr>
            <th>#</th>
            <th>Name</th>
            <th>Price</th>
            <th>Department</th>
            <th>Promo Code</th>
          </tr>
        </thead>
        <tbody>
          <ProductRows products={products} page={page} />
        </tbody>
      </Table>
    );
  }
}