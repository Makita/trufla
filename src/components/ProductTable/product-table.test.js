/* eslint-disable no-undef */
import * as React from 'react';
import ProductTable, { ProductRow, ProductRows } from './product-table';

describe('ProductRow', () => {
  it('should render correctly', () => {
    const wrapper = shallow(
      <ProductRow
        id="1"
        name="Epic Product"
        price={1.00}
        dept_name="Marketing"
        code="nopromo"
        discount={50.0}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });
});

describe('ProductRows', () => {
  it('should render correctly with no data', () => {
    const wrapper = shallow(<ProductRows products={[]} />);

    expect(wrapper).toMatchSnapshot();
  })

  it('should render correctly with data', () => {
    const products = [{
      id: 1,
      name: "Epic Product",
      price: 1.00,
      dept_name: "Marketing",
      code: "nopromo",
      discount: 50.0
    }];

    const wrapper = shallow(<ProductRows products={products} />)

    expect(wrapper).toMatchSnapshot();
  });
});

describe('ProductTable', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<ProductTable />)

    expect(wrapper).toMatchSnapshot();
  });
})