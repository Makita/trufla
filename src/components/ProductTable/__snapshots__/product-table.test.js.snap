// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductRow should render correctly 1`] = `
<tr>
  <td>
    1
  </td>
  <td>
    Epic Product
  </td>
  <td>
    <s>
      1.00
    </s>
     
    0.50
  </td>
  <td />
  <td />
</tr>
`;

exports[`ProductRows should render correctly with data 1`] = `
<ProductRow
  department="Marketing"
  discount={50}
  id={1}
  key="1"
  name="Epic Product"
  price={1}
  promoCode=""
/>
`;

exports[`ProductRows should render correctly with no data 1`] = `
<tr
  key="missing0"
>
  <td>
    #
  </td>
  <td>
    Loading
  </td>
  <td>
    Loading
  </td>
  <td>
    Loading
  </td>
  <td>
    Loading
  </td>
</tr>
`;

exports[`ProductTable should render correctly 1`] = `
<Table
  bordered={false}
  bsClass="table"
  condensed={false}
  hover={true}
  responsive={true}
  striped={true}
>
  <thead>
    <tr>
      <th>
        #
      </th>
      <th>
        Name
      </th>
      <th>
        Price
      </th>
      <th>
        Department
      </th>
      <th>
        Promo Code
      </th>
    </tr>
  </thead>
  <tbody>
    <ProductRows />
  </tbody>
</Table>
`;
